components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/dao/impl/ActionManagementDAOImpl.java

```diff
@@ -63,145 +61,134 @@ public ActionManagementDAOImpl() {
-                     ActionMgtSQLConstants.Query.GET_ACTIONS_BASIC_INFO_BY_ACTION_TYPE)) {
-
-            statement.setString(ActionMgtSQLConstants.Column.ACTION_TYPE, actionType);
-            statement.setInt(ActionMgtSQLConstants.Column.TENANT_ID, tenantId);
-
-            try (ResultSet rs = statement.executeQuery()) {
-                while (rs.next()) {
-                    String actionId = rs.getString(ActionMgtSQLConstants.Column.ACTION_UUID);
-
-                    actions.add(new Action.ActionResponseBuilder()
+        Map<String, Action.ActionResponseBuilder> actionBuilderMap = new HashMap<>();
+        NamedJdbcTemplate jdbcTemplate = new NamedJdbcTemplate(IdentityDatabaseUtil.getDataSource());
+        try {
+            jdbcTemplate.executeQuery(ActionMgtSQLConstants.Query.GET_ACTIONS_BASIC_INFO_BY_ACTION_TYPE,
+                (resultSet, rowNumber) -> {
+                    String actionId = resultSet.getString(ActionMgtSQLConstants.Column.ACTION_UUID);
+                    Action.ActionResponseBuilder builder = new Action.ActionResponseBuilder()
                             .id(actionId)
-                            .type(Action.ActionTypes.valueOf(
-                                    rs.getString(ActionMgtSQLConstants.Column.ACTION_TYPE)))
-                            .name(rs.getString(ActionMgtSQLConstants.Column.ACTION_NAME))
-                            .description(rs.getString(ActionMgtSQLConstants.Column.ACTION_DESCRIPTION))
-                            .status(Action.Status.valueOf(
-                                    rs.getString(ActionMgtSQLConstants.Column.ACTION_STATUS)))
-                            .endpoint(getActionEndpointConfigById(dbConnection, actionId, tenantId)).build());
-                }
+                            .type(Action.ActionTypes
+                                    .valueOf(resultSet.getString(ActionMgtSQLConstants.Column.ACTION_TYPE)))
+                            .name(resultSet.getString(ActionMgtSQLConstants.Column.ACTION_NAME))
+                            .description(resultSet.getString(ActionMgtSQLConstants.Column.ACTION_DESCRIPTION))
+                            .status(Action.Status
+                                .valueOf(resultSet.getString(ActionMgtSQLConstants.Column.ACTION_STATUS)));
+
+                    actionBuilderMap.put(actionId, builder);
+                    return null;
+                }, statement -> {
+                    statement.setString(ActionMgtSQLConstants.Column.ACTION_TYPE, actionType);
+                    statement.setInt(ActionMgtSQLConstants.Column.TENANT_ID, tenantId);
+                });
+
+            for (Map.Entry<String, Action.ActionResponseBuilder> entry : actionBuilderMap.entrySet()) {
+                EndpointConfig endpoint = getActionEndpointConfigById(entry.getKey(), tenantId);
+                actions.add(entry.getValue().endpoint(endpoint).build());
             }
```

Comment:
Looks like this refactoring has introduced another additional loop to go over all actions and to get the endpoint details.